engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnStateTransitionBehavior.java

```diff
@@ -154,6 +154,42 @@ public ElementInstance activateChildInstance(
     return stateBehavior.createChildElementInstance(context, childInstanceKey, childInstanceRecord);
   }
 
+  public void terminateChildInstances(final BpmnElementContext context) {
+
+    final var childInstances = stateBehavior.getChildInstances(context);
+
+    for (final BpmnElementContext childInstanceContext : childInstances) {
+
+      if (WorkflowInstanceLifecycle.canTransition(
+          childInstanceContext.getIntent(), WorkflowInstanceIntent.ELEMENT_TERMINATING)) {
+        transitionToTerminating(childInstanceContext);
+
+      } else if (childInstanceContext.getIntent() == WorkflowInstanceIntent.ELEMENT_COMPLETED) {
+        // clean up the state because the completed event will not be processed
+        stateBehavior.removeInstance(childInstanceContext);
+      }
+    }
+
+    final var elementInstance = stateBehavior.getElementInstance(context);
+    final var activeChildInstances = elementInstance.getNumberOfActiveElementInstances();
+
+    if (activeChildInstances == 0) {
+      // terminate element instance if all child instances are terminated
+      transitionToTerminated(context);
+
+    } else {
+      // wait for child instances to be terminated
```

Comment:
Nope. In this case, the element instance has child instances that are not terminated yet. So, it should wait until all child instances are terminated before it goes to `terminated`.

Does this make sense for you?

Reply:
Yes, but the code does not wait, it merely consumes the pending tokens. Or am I missing something?

Reply:
Ok, I guess that my comment is not clear :sweat_smile:  I'm open to suggestions.

"waiting" here means that the element instance doesn't go to `terminated`. Instead, it waits until the child instances are terminated.

It doesn't consume the tokens from the child instances :thinking: It consumes the tokens for the events that will not be processed anymore:
* child instance `completed` doesn't pass the guard because the (flow scope) instance is not active
* `sequence flow taken` doesn't pass the guard because the (flow scope) instance is not active 

For both cases, the (flow scope) instance has a pending token. So, we need to "consume" these tokens to pass the check: https://github.com/zeebe-io/zeebe/blob/13a78a6fda2ba73c886c1d3f3087c48f18f0072e/engine/src/main/java/io/zeebe/engine/nwe/container/MultiInstanceBodyProcessor.java#L210-L228

Reply:
Aaah! Okay, so we juts don't transition to terminated yet. For me, it would have helped to just add this intent to the comment. So something like: 
```suggestion
      // don't yet transition to terminated, but wait for child instances to be terminated
```